### ISSUE TITLE
Rename `version.log` to `version.txt`

### ISSUE THREAD

**Issue by gautam-kotian-sociomantic** [sociomantic-tsunami/ocean#517]
`version.log` is different from other application logs, as it is only written to a single time when an application starts, and is then immediately closed. For this reason, an application doesn't need to reopen it after a log rotation. Moreover, to be GDPR compliant, applications generally use a logrotate configuration of the form `perform rotations weekly + keep about 10 previously rotated log files + delete logs older than 60 days`. This combination results in a not-so-long history in the `version.log`.

To get around these problems, I propose that `version.log` be renamed to `version.txt` (i.e. change the default value of the `default_file` argument in [this line](https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/util/app/ext/VersionArgsExt.d#L127)).

Doing this will bring the following benefits:
- clear separation of `version.txt` from other application log files
- logrotate configuration of `version.txt` can be specified individually without having to resort to complicated file-globbing mechanisms
- code that registers `version.txt` with the ReopenableFilesExt can be removed
- there will be no need to send any reopen signal to the application when `version.txt` is rotated (simply use `copytruncate` in the logrotate configuration)

**Comment by nemanja-boric-sociomantic** [comment]
> there will be no need to send any reopen signal to the application when version.txt (simply use `copytruncate` in the logrotate configuration)

I'm curious why `copytruncate` is needed for this (it will work just fine even without it - just rename and touch new file is fine, just don't do anything in the `postrotate`, AFAICS)?

**Comment by gautam-kotian-sociomantic** [comment]
>I'm curious why copytruncate is needed for this?

Right. Even that wouldn't be needed.

If we go ahead with this, I foresee logrotate configurations of applications to be of the form:
```
/path/to/applications/log/directory/*.log
{
    ...
}

/path/to/applications/log/directory/version.txt
{
    ...
}
```

**Comment by nemanja-boric-sociomantic** [comment]
I'm curious to hear what @leandro-lucarella-sociomantic says about it. I feel a bit uneasy when it comes to naming log files `*.txt`, but maybe it's worth it, especially because it already ends up in the log directory.

**Comment by gavin-norman-sociomantic** [comment]
>I feel a bit uneasy when it comes to naming log files *.txt

Right. Feels like a hack to me.

Maybe we should address the question: what is the need for keeping version logs beyond 60 days?

**Comment by leandro-lucarella-sociomantic** [comment]
I still fail a bit to see why is it problematic to keep `version.log` as it is, just for consistency (and avoid special casing in configs, specially rolling out new configs for logrotate is painful). Did you ever needed to check which version was ran more then 2-3 months ago?

I ACK that the version.log is a very clumsy mechanism for what we need, but hacking it to be even more obscure doesn't seem like a good approach (I would rather see graphite events to be emitted each time an application start, stating the version, for example).

Incidentally (or maybe not), we were discussing the version log recently with @gavin-norman-sociomantic too (maybe it was in the leads meeting? :D).

**Comment by nemanja-boric-sociomantic** [comment]
I can see the issue where the application has uptime of more than 60 days. Then you pretty much loose all the history of the past versions (if ever needed). In the past we had larger issue where you also loosed the information about the current version (if the binary was replaced in the meantime), that's why I added support to query the version log from the running application via unix socket: #478

**Comment by joseph-wakeling-sociomantic** [comment]
I don't think this is a correct solution.  It's a workaround, it doesn't solve the problem.

The correct thing to do is to have logrotate config which appropriately special-cases `version.log` as the app maintainer needs.

Question: won't it work to do something like:

```
/path/to/appname/log/*.log
{
    rotate 10
    weekly
    maxage 60
    missingok
    notifempty
    delaycompress
    compress
    maxsize 100M
    sharedscripts
    postrotate
        service appname reload
    endscript
}

/path/to/appname/log/version.log
{
    size 100M
}
```

... as the later rule will just override the one part of the more general rules defined above?  Note, the above just switches the rotation condition from a time-based one (`weekly`) to a size-based one (`size 100M`).

**Comment by nemanja-boric-sociomantic** [comment]
Don't take my word for it, but:

> The developers of logrotate have intentionally forbidden overlapping log-file definitions, so whilst more-specific rules are allowed to follow general rules in the configuration, log-file path overlaps still generate an error. This is partly for the benefit of package-maintainers who include logrotate rules, and the users of those packages, so that overlaps become obvious and can be reported.

https://superuser.com/questions/913579/logrotate-how-to-configure-and-override-defaults

So, no, that doesn't work.

**Comment by joseph-wakeling-sociomantic** [comment]
Damn, that's annoying :-(  I can confirm having tried it out that it doesn't work.

**Comment by nemanja-boric-sociomantic** [comment]
Since more and more applications are now hit by this, we should consider if we want to continue with this PR. It is a hack, but all other options are much worse IMHO (except unix socket approach, but that doesn't keep the history).

**Comment by joseph-wakeling-sociomantic** [comment]
Yea, and it'll get worse and worse as more apps switch to using `ifempty`.  Are there any longer-term plans for how we're going to manage logs?  Otherwise this seems like a reasonable compromise.

**Comment by joseph-wakeling-sociomantic** [comment]
Just for the sake of my poor and failing memory ... what ARE the other options?

For example, if it's possible for apps to specify a different output filename in config, that might be preferable to hardcoding a hack like this.

**Comment by nemanja-boric-sociomantic** [comment]
I'm discussing with the infrastructure if they can keep the version history for us, since we're in the long term moving away from the log _files_ (they will be managed by infrastructure, if any).

**Comment by nemanja-boric-sociomantic** [comment]
> Just for the sake of my poor and failing memory ... what ARE the other options?

All the options mentioned here, that one hasn't . But then all applications will have to have

```
[LOG.ocean.util.app.ext.VersionArgsExt]
file=log/version.txt
```

which I'm not sure how better it is than doing that for all the applications. But since this is a breaking change and we can move with it only in ocean v5.0.0, most of the applications will do the same anyway.

**Comment by joseph-wakeling-sociomantic** [comment]
Doing it via config avoids the breaking change and makes it possible to limit the change only to apps/maintainers who are bothered by existing behaviour.  Given that we have a long-term change in mind with infrastructure, that seems the least intrusive option.

**Comment by gavin-norman-sociomantic** [comment]
>I'm discussing with the infrastructure if they can keep the version history for us, since we're in the long term moving away from the log files (they will be managed by infrastructure, if any).

This would be the ideal. We _almost_ have all the info already.
